src/IECoreScene/SceneCache.cpp

```diff
@@ -728,8 +730,76 @@ class SceneCache::ReaderImplementation : public SceneCache::Implementation
 			return reader;
 		}
 
+		ConstPathMatcherDataPtr readSet( const Name &name ) const
+		{
+			IECore::PathMatcherDataPtr pathMatcherData = readLocalSet( name );
+
+			NameList children;
+			childNames( children );
+
+			for( const auto &c : children )
+			{
+				ReaderImplementation *nc = const_cast<ReaderImplementation *>(this);
+				ConstPathMatcherDataPtr childSets = nc->child( c, SceneInterface::NullIfMissing )->readSet( name );
+				pathMatcherData->writable().addPaths(childSets->readable(), { c });
+			}
+
+			return pathMatcherData;
+		}
+
+		NameList setNames() const
+		{
+			NameList setNames;
+			IndexedIOPtr setsIO = m_indexedIO->subdirectory( setsEntry, IECore::IndexedIO::NullIfMissing );
+			if( setsIO )
+			{
+				IndexedIO::EntryIDList localSetNames;
+				setsIO->entryIds(localSetNames, IndexedIO::Directory);
+
+				for ( auto localSetName : localSetNames)
+				{
+					setNames.push_back( localSetName );
+				}
+			}
+
+			NameList children;
+			childNames( children );
+
+			for (const auto &c : children)
+			{
+				ReaderImplementation* nc = const_cast<ReaderImplementation*>(this);
+				NameList childSetNames = nc->child(c, SceneInterface::NullIfMissing)->setNames();
+				for (const auto csn : childSetNames)
+				{
+					setNames.push_back(csn);
+				}
+			}
+
+			std::set<SceneInterface::Name> unique(setNames.begin(), setNames.end());
```

Comment:
Perhaps we should add a typedef set<SceneInterface::Name> NameSet to SceneInterface? What do you think to updating the setNames return to NameSet?

Reply:
My guess is that using `std::set` to make things unique is going to be a fair bit slower than a combo of `std::sort` and `std::unique` done in-place on the vector itself. I can see the logic in using a datatype that enforces uniqueness for the return value, but I suspect this is of less practical use to the caller than performance is.

Reply:
How do you feel about updating functions which return containers as reference parameters to functions which directly return them?

Reply:
I think I'd like that, but I'd also leave it out of the sets work. Let's just focus on the sets for now. Maybe there's even an argument for making `setNames()` be pass-by-reference and then updating everything all at once later?

And I guess pass-by-reference is still better if I happen to have a suitable container already, for instance if I had a list of locations to iterate over, and at each step needed to call `setNames()`. I could just make a container outside the loop and then reuse it for each step.